proving.go

```diff
@@ -0,0 +1,270 @@
+		if !found {
+			// If there are no more leaves to prove - we're done.
+			break
+		}
+
+		additionalProof, currentPos, err := cache.calcProofForNextLeaf(nextProvenLeafPos, provenLeafIndexIt)
+		if err != nil {
+			return nil, err
+		}
+		proof = append(proof, additionalProof...)
+
+		for ; currentPos.height < rootHeight; currentPos = currentPos.parent() { // Traverse cache:
+
+			// Check if we're revisiting a node. If we've descended into a subtree and just got back, we shouldn't add
+			// the sibling to the proof and instead move on to the parent.
+			found := skipPositions.PopIfEqual(currentPos)
+			if found {
+				continue
+			}
+
+			// If the current node sibling is an ancestor of the next proven leaf sibling we should process it's subtree
+			// instead of adding it to the proof. When we reach it again we'll want to skip it.
+			if p, found := provenLeafIndexIt.peek(); found && currentPos.sibling().isAncestorOf(p) {
+				skipPositions.Push(currentPos.sibling())
+				break
+			}
+
+			currentVal, err := cache.GetNode(currentPos.sibling())
+			if err != nil {
+				return nil, err
+			}
+
+			proof = append(proof, currentVal)
+		}
+	}
+
+	return proof, nil
+}
+
+type TreeCache struct {
+	readers map[uint]NodeReader
+	hash    HashFunc
+}
+
+// GetNode reads the node at the requested position from the cache or calculates it if not available.
+func (c *TreeCache) GetNode(nodePos position) ([]byte, error) {
+	// Get the cache reader for the requested node's layer.
+	reader, found := c.readers[nodePos.height]
+
+	// If the cache wan't found, we calculate the minimal subtree that will get us the required node.
+	if !found {
+		return c.calcNode(nodePos)
+	}
+
+	err := reader.Seek(nodePos.index)
+	if err != nil {
+		return nil, errors.New("while seeking in cache: " + err.Error() + nodePos.String())
+	}
+	currentVal, err := reader.ReadNext()
+	if err != nil {
+		return nil, errors.New("while reading from cache: " + err.Error())
+	}
+	return currentVal, nil
+}
+
+func (c *TreeCache) calcNode(nodePos position) ([]byte, error) {
+	var subtreeStart position
+	var found bool
+	var reader NodeReader
+
+	// Find the next cached layer below the current one.
+	for subtreeStart = nodePos.leftChild(); !found; subtreeStart = subtreeStart.leftChild() {
+		reader, found = c.readers[subtreeStart.height]
+	}
+
+	// Prepare the reader for traversing the subtree.
+	err := reader.Seek(subtreeStart.index)
+	if err != nil {
+		return nil, errors.New("while seeking in cache: " + err.Error() + subtreeStart.String())
+	}
+
+	// Traverse the subtree.
+	width := uint64(1) << (nodePos.height - subtreeStart.height)
+	_, currentVal, err := traverseSubtree(reader, width, c.hash, nil)
+	if err != nil {
+		return nil, errors.New("while traversing subtree for root: " + err.Error())
+	}
+	return currentVal, nil
+}
+
+func (c *TreeCache) calcProofForNextLeaf(nextProvenLeafPos position, provenLeafIndexIt *positionsIterator) ([][]byte,
```

Comment:
this method shouldn't be under `TreeCache`. It uses the cache but it's far from serving a cache logic.. Make this method a function and rename it to `calcProofForLeaf` (the "next" is a context known only to the caller, the function doesn't care about that fact).